src/cp/crystallography.cxx

```diff
@@ -0,0 +1,467 @@
+  }
+
+  return res;
+}
+
+SymmetryGroup::SymmetryGroup(std::string sclass) :
+    ops_(symmetry_rotations(sclass))
+{
+  raw_.resize(ops_.size()*4);
+  size_t j = 0;
+  for (auto it = ops_.begin(); it != ops_.end(); ++it) {
+    std::copy(it->quat(), it->quat()+4, &raw_[j]);
+    j += 4;
+  }
+}
+
+SymmetryGroup::~SymmetryGroup()
+{
+
+}
+
+std::string SymmetryGroup::type()
+{
+  return "SymmetryGroup";
+}
+
+ParameterSet SymmetryGroup::parameters()
+{
+  ParameterSet pset(SymmetryGroup::type());
+
+  pset.add_parameter<std::string>("sclass");
+
+  return pset;
+}
+
+std::unique_ptr<NEMLObject> SymmetryGroup::initialize(ParameterSet & params)
+{
+  return neml::make_unique<SymmetryGroup>(
+      params.get_parameter<std::string>("sclass")); 
+}
+
+const std::vector<Orientation> & SymmetryGroup::ops() const
+{
+  return ops_;
+}
+
+size_t SymmetryGroup::nops() const
+{
+  return ops_.size();
+}
+
+Orientation SymmetryGroup::misorientation(const Orientation & a, 
+                                          const Orientation & b) const
+{
+  Orientation best;
+  double angle_best = 2 * M_PI;
+  double cn[3];
+  double ca;
+  for (auto ri = ops_.begin(); ri != ops_.end(); ++ri) {
+    for (auto rj = ops_.begin(); rj != ops_.end(); ++rj) {
+      Orientation trial = *ri * a * b.inverse() * *rj;
+      trial.to_axis_angle(cn, ca);
+      if (ca < angle_best) {
+        angle_best = ca;
+        best = trial;
+      }
+    }
+  }
+
+  return best;
+}
+
+double SymmetryGroup::distance(const Orientation & a, const Orientation & b)
+  const
+{
+  double * dd = new double [nops()];
+  
+  opdist_(a.quat(), b.quat(), dd);
+
+  double val = *std::min_element(&dd[0], &dd[nops()]);
+
+  delete [] dd;
+
+  return val;
+}
+
+void SymmetryGroup::closest(const Orientation & a, const Orientation & b, 
+                            Orientation & close, double & dist) const
+{
+  double * dd = new double [nops()];
+  
+  opdist_(a.quat(), b.quat(), dd);
+
+  double * loc = std::min_element(&dd[0], &dd[nops()]);
+
+  dist = *loc;
+
+  size_t ind = loc - &dd[0];
+
+  delete [] dd;
+
+  close = ops_[ind] * a;
+}
+
+void SymmetryGroup::opdist_(const double * const q1, 
+                            const double * const q2, double * const res) 
+  const
+{
+  double * qs = new double [4 * nops()];
+  
+  qmult_vec(&raw_[0], q1, nops(), qs);
+  dgemv_("T", 4, nops(), 1.0, qs, 4, q2, 1, 0.0, res, 1);
+
+  delete [] qs;
+
+  for (size_t i=0; i<nops(); i++) {
+    res[i] = acos(fabs(res[i]));
+  }
+}
+
+Lattice::Lattice(Vector a1, Vector a2, Vector a3,
+                 std::shared_ptr<SymmetryGroup> symmetry,
+                 list_systems isystems) :
+    a1_(a1), a2_(a2), a3_(a3), symmetry_(symmetry), offsets_({0}),
+    setup_(false)
+{
+  make_reciprocal_lattice_();
+
+  for (auto system : isystems) {
+    add_slip_system(system.first, system.second);
+  }
+}
+
+Lattice::~Lattice()
+{
+
+}
+
+Vector Lattice::miller2cart_direction(std::vector<int> m)
```

Comment:
does this take into account hexagonal 4 index system ?

Reply:
No, you would have to convert to Miller indices first.  If we want HCP I can easily add that.  My old LLNL would do the conversions for you automatically.  For now I'm not going to worry about it as I don't see us doing any HCP metals sooner.